Add NEX to coins.

modenero opened the pull request:
**NEXA** is an all-new Proof-of-Work, UTXO blockchain that was recently launched on June 21st.

For project details, please visit:
https://www.nexa.org/

For block explorer, please visit:
https://explorer.nexa.org/

For social media, please visit:
https://www.reddit.com/r/Nexa/

cipig commented:
Would like to see a swap with this coin, idk if it works.
Something like https://github.com/KomodoPlatform/coins/blob/master/swaps/LCC_KMD.md

modenero commented:
@cipig 
@smk762 

thanks so much for providing such clear guidance for NEX integration to the platform. I've spent a few days working on this but I'm not making much progress towards testing a KMD/NEX exchange.

i'm attempting to build atomicDEX-API using `cargo build -vv`, but I'm getting the following error:

```
[zstd-sys 1.6.2+zstd.1.5.1] exit status: 0
[zstd-sys 1.6.2+zstd.1.5.1] running: "ar" "s" "/home/dev/Workspace/atomicDEX-API/target/debug/build/zstd-sys-54c412f651ffb57e/out/libzstd.a"
[zstd-sys 1.6.2+zstd.1.5.1] exit status: 0
[zstd-sys 1.6.2+zstd.1.5.1] cargo:rustc-link-lib=static=zstd
[zstd-sys 1.6.2+zstd.1.5.1] cargo:rustc-link-search=native=/home/dev/Workspace/atomicDEX-API/target/debug/build/zstd-sys-54c412f651ffb57e/out
[zstd-sys 1.6.2+zstd.1.5.1] cargo:root=/home/dev/Workspace/atomicDEX-API/target/debug/build/zstd-sys-54c412f651ffb57e/out
[zstd-sys 1.6.2+zstd.1.5.1] cargo:include=/home/dev/.cargo/registry/src/github.com-1ecc6299db9ec823/zstd-sys-1.6.2+zstd.1.5.1/zstd/lib
error: build failed
```

__I'm running Ubuntu 22.04 in a virtual env__

---

unfortunately, i don't have access to discord or tg, so if you wouldn't mind offering your support here in github, that would be very much appreciated.

thanks again!

smk762 commented:
Do the release binaries work for you? https://github.com/KomodoPlatform/atomicDEX-API/releases/tag/beta-2.1.7404

smk762 commented:
Here's an example MM2.json which should get you going - 
```json
{
    "gui": "MM2CLI",
    "netid": 7777,
    "rpc_password": "RPC_CoN7RoL_U53RP@55W@RD",
    "passphrase": "ENTER SEED PHASE OR PRIVATE KEY HERE",
    "userhome": "/${HOME#\"/\"}"
}
```
Coins file should to be in same folder as mm2 binary. Was NEX source code forked from a coin? Decimals value looks low. Is address same for NEX and BTC or BCH using same seed? Need to confirm the base58 params in the source code. 
If you can link me to the NEX github repository, I"ll have a look.

cipig commented:
> KMD is definitly present in the `coins` file that I copied from:
> https://github.com/jl777/coins/blob/master/coins

Please use https://github.com/KomodoPlatform/coins/blob/master/coins
You need to enable the coins first before you can work with them, eg:
```
curl --url "http://127.0.0.1:7783" --data "{\"method\":\"electrum\",\"coin\":\"KMD\",\"servers\":[{\"url\":\"electrum1.cipig.net:20001\",\"protocol\":\"SSL\"},{\"url\":\"electrum2.cipig.net:10001\"},{\"url\":\"electrum3.cipig.net:10001\"}],\"userpass\":\"$userpass\"}"
```

modenero commented:
@smk762 
@cipig 

spent the weekend with AtomicDEX docs and made some terrific headway. so impressed by what your team is doing with atomic "cross-chain" swaps 👏 👍 was able to successfully trade `RICK` and `MORTY` via my locally running API. and i'm looking forward to working next with the Market Maker as well as your WASM implementation (for our web-based exchange).

unfortunately though, i wasn't able to complete the testing for `NEX`. when I tried to add it, I received this error:
```
{
  "error": "rpc:212] dispatcher_legacy:160] lp_commands_legacy:102] lp_coins:2310] utxo_standard:53] utxo_coin_builder:448] Failed to connect to at least 1 of [ElectrumRpcRequest { url: \"electrum.nexa.org:20001\", protocol: TCP, disable_cert_verification: false }] in 5 seconds."
}
```

my `NEXconnect.sh` file contains:
```
#!/bin/bash
source userpass
curl --url "http://127.0.0.1:7783" --data "{\"userpass\":\"$userpass\",\"method\":\"electrum\",\"coin\":\"NEX\",\"servers\":[{\"url\":\"electrum.nexa.org:20001\"}]}" | jq
```

_(NOTE: we're now working on setting up a 2nd "public" electrum server before officially listing this new coin)_

`electrum.nexa.org:20001` is actively being used for https://explorer.nexa.org/, so i'm confident it should be working properly.

---

@smk762 in response to your questions:

Was NEX source code forked from a coin? 
yes. Bitcoin Cash (BCH)

Decimals value looks low. 
decimals have been moved 6 places to eliminate the "decimals problem" when transacting. 1 MEX = 1,000,000 NEX. and 1 MEX would be the equivalent of 1 BTC or 1 BCH in terms of the economics of this coin.

Is address same for NEX and BTC or BCH using same seed? 
i'm not sure. i'm working on that now and should be able to update you soon

Need to confirm the base58 params in the source code. If you can link me to the NEX github repository, I"ll have a look.
https://gitlab.com/nexa/nexa

---

very much looking forward to integrating Nexa with the AtomicDEX and really appreciate your time and patience in helping me do so.

cheers!

cipig commented:
there is nothing listening on port `20001` on `electrum.nexa.org`
likely wrong port
you can also check by sending commands to the server, eg with `(echo '{ "id": 1, "method": "server.version", "params": ["AtomicDEX", ["1.4", "2.0"]] }'; sleep 0.5) | ncat electrum.nexa.org 20001` to connect, call server.version and get the response

smk762 commented:
Base58 params appear to be incorrect in your PR - https://gitlab.com/nexa/nexa/-/blob/dev/src/chainparams.cpp#L580-582

modenero commented:
> Base58 params appear to be incorrect in your PR - https://gitlab.com/nexa/nexa/-/blob/dev/src/chainparams.cpp#L580-582

ahh yes, i see that now 👀
thank you! 🙏

i should have compared it to BitcoinCashNode (https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/blob/master/src/chainparams.cpp#L210)

still much for me to learn and understand at this low-level

anyway, i've updated the values and "force-" pushed the update

---

@smk762
i also updated `"bech32_hrp": "nq"` to `"bech32_hrp": "nexa"` now that i know what `hrp` stands for. 

my question is that the BCH entry in `coins` uses `address_format` and NOT `bech32_hrp`. so should NEX use one over the other? or is it okay to have both entries?

---

and btw i'm still working on resolving the Electrum server issues...

smk762 commented:
> i also updated "bech32_hrp": "nq" to "bech32_hrp": "nexa" now that i know what hrp stands for.
> my question is that the BCH entry in coins uses address_format and NOT bech32_hrp. so should NEX use one over the other? or is it okay to have both entries?

The `addressformat.network` bit would be the first bit, the "cashaddress" address prefix, and the `bech32_hrp` would be the first 2 letters of the actual address.

So assuming an address like `nexa:nqtsq5g5ja7ukvt2ftehmun9gnjrkxesp7hudjhek9tnm5m5` 

`addressformat.network`  = "nexa"

 `bech32_hrp` = "nq"

modenero commented:
> `bech32_hrp` = "nq"

ahhh, okay that makes sense now :+1: 

---
@smk762 
just want to thank you so much for your time and patience :pray: 

so we've fixed the port issue on the primary Electrum server and added a second server .. both are working now

i'm able to successfully activate NEX and deposit to the address
https://explorer.nexa.org/address/nexa:qqwnm809af5muc4cjua74d7ze3lwqt0utvmx7u8kyg

the balance reports correctly for `./mybalance.sh NEX`:
```
{
  "coin": "NEX",
  "balance": "13370",
  "unspendable_balance": "0",
  "address": "nexa:qqwnm809af5muc4cjua74d7ze3lwqt0utvmx7u8kyg"
}
```

unfortunately, when I try to withdraw, the transaction format generated by the API is not valid .. it will decode perfectly for BTC and BCH, so I know its working as expected, but something needs to be changed with the transaction format for NEX.

i'm working with the Nexa team to sort out how to best complete this DEX integration. they offered 2 suggestions:
1. integrate `libnexa.so` into atomicDEX
2. use the `nexa-tx` program

i'm not sure if these are viable options, so I would like to ask you if you can recommend a solution, based on your past integration experience?

here is a link to the Nexa transaction primitive
- https://gitlab.com/nexa/nexa/-/blob/dev/src/primitives/transaction.cpp

and here are 2 links to Nexa protocol specs
- https://spec.nexa.org/protocol/blockchain/transaction.md
- https://spec.nexa.org/nexa/transactionIdentifier.md

and here is the link to the Nexa tx decoder
- https://explorer.nexa.org/decoder

any advice is very much appreciated

modenero commented:
@smk762 
thanks for the advice and all the support :pray: 

@artemii235 
really appreciate your assistance for this integration
i'll close this issue now and continue in the new issue
